utils/sparse.py

```diff
@@ -0,0 +1,151 @@
+import math
+
+from sparsezoo import Zoo
+from sparseml.pytorch.optim import ScheduledModifierManager
+from sparseml.pytorch.utils import SparsificationGroupLogger
+
+from utils.torch_utils import is_parallel
+
+
+def _get_model_framework_file(model, path):
+    transfer_request = 'recipe_type=transfer' in path
+    checkpoint_available = any([file.checkpoint for file in model.framework_files])
+    final_available = any([not file.checkpoint for file in model.framework_files])
+
+    if transfer_request and checkpoint_available:
+        # checkpoints are saved for transfer learning use cases,
+        # return checkpoint if avaiable and requested
+        return [file for file in model.framework_files if file.checkpoint][0]
+    elif final_available:
+        # default to returning final state, if available
+        return [file for file in model.framework_files if not file.checkpoint][0]
+
+    raise ValueError(f"Could not find a valid framework file for {path}")
+
+
+def check_download_sparsezoo_weights(path):
+    if isinstance(path, str):
+        if path.startswith("zoo:"):
+            # load model from the SparseZoo and override the path with the new download
+            model = Zoo.load_model_from_stub(path)
+            file = _get_model_framework_file(model, path)
+            path = file.downloaded_path()
+
+        return path
+
+    if isinstance(path, list):
+        return [check_download_sparsezoo_weights(p) for p in path]
+
+    return path
+
+
+class SparseMLWrapper(object):
+    def __init__(self, model, checkpoint_recipe, train_recipe):
+        self.enabled = bool(checkpoint_recipe or train_recipe)
+        self.model = model.module if is_parallel(model) else model
+        self.checkpoint_manager = ScheduledModifierManager.from_yaml(checkpoint_recipe) if checkpoint_recipe else None
+        self.manager = ScheduledModifierManager.from_yaml(train_recipe) if train_recipe else None
+        self.logger = None
+        self.start_epoch = None
+
+    def state_dict(self):
+        if self.checkpoint_manager:
+            manager = ScheduledModifierManager.compose_staged(self.checkpoint_manager, self.manager)
+        else:
+            manager = self.manager
+        return {
+            'recipe': str(manager) if self.enabled else None,
+        }
+
+    def apply(self, epoch):
```

Comment:
let's rename to `apply_checkpoint_structure` for clarity on apply here since it shares a name with the manager class (and that could ostensibly refer to the train manager as well)